// simple-plist/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeFileSync = exports.writeFile = exports.writeBinaryFileSync = exports.writeBinaryFile = exports.stringify = exports.readFileSync = exports.readFile = exports.parse = exports.bplistParser = exports.bplistCreator = void 0;
var bplist_creator_1 = require("bplist-creator");
Object.defineProperty(exports, "bplistCreator", { enumerable: true, get: function () { return __importDefault(bplist_creator_1).default; } });
var bplist_parser_1 = require("bplist-parser");
Object.defineProperty(exports, "bplistParser", { enumerable: true, get: function () { return __importDefault(bplist_parser_1).default; } });
var parse_1 = require("./parse");
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return parse_1.parse; } });
var readFile_1 = require("./readFile");
Object.defineProperty(exports, "readFile", { enumerable: true, get: function () { return readFile_1.readFile; } });
var readFileSync_1 = require("./readFileSync");
Object.defineProperty(exports, "readFileSync", { enumerable: true, get: function () { return readFileSync_1.readFileSync; } });
var stringify_1 = require("./stringify");
Object.defineProperty(exports, "stringify", { enumerable: true, get: function () { return stringify_1.stringify; } });
var writeBinaryFile_1 = require("./writeBinaryFile");
Object.defineProperty(exports, "writeBinaryFile", { enumerable: true, get: function () { return writeBinaryFile_1.writeBinaryFile; } });
var writeBinaryFileSync_1 = require("./writeBinaryFileSync");
Object.defineProperty(exports, "writeBinaryFileSync", { enumerable: true, get: function () { return writeBinaryFileSync_1.writeBinaryFileSync; } });
var writeFile_1 = require("./writeFile");
Object.defineProperty(exports, "writeFile", { enumerable: true, get: function () { return writeFile_1.writeFile; } });
var writeFileSync_1 = require("./writeFileSync");
Object.defineProperty(exports, "writeFileSync", { enumerable: true, get: function () { return writeFileSync_1.writeFileSync; } });
